File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.py
def solution(n, words):
    used = set()
    used.add(words[0])
    prev = words[0][-1]
    cnt = 1

    for word in words[1:]:
        if word in used or prev != word[0]:
            return cnt % n + 1, cnt // n + 1

        used.add(word)
        cnt += 1
        prev = word[-1]

    return 0, 0

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.HashSet;
import java.util.Set;

class Solution {

    public int[] solution(int n, String[] words) {
        Set<String> usedWords = new HashSet<>();
        char previousLastChar = '\0';

        for (int i = 0; i < words.length; i++) {
            String word = words[i];

            boolean disconnected =
                    i > 0 && previousLastChar != word.charAt(0);

            boolean duplicated = !usedWords.add(word);

            if (disconnected || duplicated) {
                int player = i % n + 1;
                int turn = i / n + 1;

                return new int[]{player, turn};
            }

            previousLastChar = word.charAt(word.length() - 1);
        }

        return new int[]{0, 0};
    }
}